lib/scholar/neighbors/random_projection_forest.ex

```diff
@@ -0,0 +1,363 @@
+          ]
+        ]
+      >
+  """
+  deftransform predict(forest, x) do
+    if Nx.rank(x) != 2 do
+      raise ArgumentError,
+            """
+            expected input tensor to have shape {num_samples, num_features}, \
+            got tensor with shape: #{inspect(Nx.shape(x))}\
+            """
+    end
+
+    if Nx.axis_size(forest.hyperplanes, 2) != Nx.axis_size(x, 1) do
+      raise ArgumentError,
+            """
+            expected hyperplanes and input tensor to have the same dimension, \
+            got #{inspect(Nx.axis_size(forest.hyperplanes, 2))} \
+            and #{inspect(Nx.axis_size(x, 1))}
+            """
+    end
+
+    predict_n(forest, x)
+  end
+
+  defn predict_n(forest, x) do
+    num_trees = forest.num_trees
+    leaf_size = forest.leaf_size
+    indices = forest.indices |> Nx.vectorize(:trees)
+    start_indices = compute_start_indices(forest, x, leaf_size: leaf_size) |> Nx.new_axis(1)
+    size = Nx.axis_size(x, 0)
+
+    pos =
+      Nx.iota({1, 1, leaf_size})
+      |> Nx.broadcast({num_trees, size, leaf_size})
+      |> Nx.vectorize(:trees)
+      |> Nx.add(start_indices)
+
+    Nx.take(indices, pos)
+    |> Nx.devectorize()
+    |> Nx.rename(nil)
+    |> Nx.transpose(axes: [1, 0, 2])
+  end
+
+  defn compute_start_indices(forest, x, opts) do
+    leaf_size = opts[:leaf_size]
+    size = Nx.axis_size(x, 0)
+    depth = forest.depth
+    num_trees = forest.num_trees
+    hyperplanes = forest.hyperplanes |> Nx.vectorize(:trees)
+    medians = forest.medians |> Nx.vectorize(:trees)
+
+    {start_indices, left?, cell_sizes, _} =
+      while {
+              start_indices = Nx.broadcast(Nx.u32(0), {num_trees, size}) |> Nx.vectorize(:trees),
+              _left? = Nx.broadcast(Nx.u8(0), {num_trees, size}) |> Nx.vectorize(:trees),
+              cell_sizes = Nx.broadcast(Nx.u32(size), {num_trees, size}) |> Nx.vectorize(:trees),
+              {
+                x,
+                hyperplanes,
+                medians,
+                level = 0,
+                nodes = Nx.broadcast(Nx.u32(0), {num_trees, size}) |> Nx.vectorize(:trees)
+              }
+            },
+            level < depth do
+        h = hyperplanes[level]
+        median = Nx.take(medians, nodes)
+        proj = Nx.dot(x, h)
+        left? = proj <= median
+
+        nodes =
+          Nx.select(
+            left?,
+            left_child(nodes),
+            right_child(nodes)
+          )
+
+        right_sizes = Nx.quotient(cell_sizes, 2)
+        left_sizes = right_sizes + Nx.remainder(cell_sizes, 2)
+        start_indices = Nx.select(left?, start_indices, start_indices + left_sizes)
+        cell_sizes = Nx.select(left?, left_sizes, right_sizes)
+
+        {
+          start_indices,
+          left?,
+          cell_sizes,
+          {x, hyperplanes, medians, level + 1, nodes}
+        }
+      end
+
+    Nx.select(not left? and cell_sizes < leaf_size, start_indices - 1, start_indices)
+  end
+
+  defn left_child(nodes) do
```

Comment:
We can make them private here if you are not expecting people to navigate this tree (I think it is harder to navigate than the kd-tree ones, no?)

Reply:
Agreed.